crates/bevy_scene/src/serde.rs

```diff
@@ -535,6 +557,57 @@ mod tests {
         assert_eq!(1, dst_world.query::<&Baz>().iter(&dst_world).count());
     }
 
+    #[test]
+    fn should_roundtrip_with_later_generations_and_obsolete_references() {
+        let mut world = create_world();
+
+        world.spawn_empty().despawn();
+
+        let a = world.spawn_empty().id();
+        let foo = world.spawn(MyEntityRef(a)).insert(Foo(123)).id();
+        world.despawn(a);
+        world.spawn(MyEntityRef(foo)).insert(Bar(123));
+
+        let registry = world.resource::<AppTypeRegistry>();
+
+        let scene = DynamicScene::from_world(&world, registry);
+
+        let serialized = scene
+            .serialize_ron(&world.resource::<AppTypeRegistry>().0)
+            .unwrap();
+        let mut deserializer = ron::de::Deserializer::from_str(&serialized).unwrap();
+        let scene_deserializer = SceneDeserializer {
+            type_registry: &registry.0.read(),
+        };
+
+        let deserialized_scene = scene_deserializer.deserialize(&mut deserializer).unwrap();
+
+        let mut map = EntityMap::default();
+        let mut dst_world = create_world();
+        deserialized_scene
+            .write_to_world(&mut dst_world, &mut map)
+            .unwrap();
+
+        assert_eq!(2, deserialized_scene.entities.len());
+        assert_scene_eq(&scene, &deserialized_scene);
+
+        let bar_to_foo = dst_world
+            .query_filtered::<&MyEntityRef, Without<Foo>>()
+            .get_single(&dst_world)
+            .unwrap()
+            .0;
+        let foo = dst_world
+            .query_filtered::<Entity, With<Foo>>()
+            .get_single(&dst_world)
+            .unwrap();
+
+        assert_eq!(foo, bar_to_foo);
```

Comment:
Could you do the `bar_to_foo.0` in the `assert_eq!` rather than when assigning to `bar_to_foo`? I think it makes the test a bit easier to understand, as the type structure is easier to parse at a glance.